ProjFS.Linux/PrjFSLib.Linux.Managed/VirtualizationInstance.cs

```diff
@@ -61,19 +61,35 @@ public class VirtualizationInstance
                 HandlePermEvent = this.preventGCOnPermEventDelegate = new ProjFS.EventHandler(this.HandlePermEvent)
             };
 
-            ProjFS fs = ProjFS.New(
+            // determine whether storageRootFullPath contains only .git and .gitattributes
+
+            string[] args;
+            if (IsUninitializedMount(storageRootFullPath))
+            {
+                args = new string[] { "-o", "initial" };
+            }
+            else
+            {
+                args = new string[] { };
+            }
+
+            this.projfs = ProjFS.New(
                 storageRootFullPath,
                 virtualizationRootFullPath,
-                handlers);
+                handlers,
+                args);
 
-            if (fs == null)
+            this.virtualizationRoot = virtualizationRootFullPath;
```

Comment:
It sounds like we can be racing with ourselves, here, in a sense -- a second process may try to enumerate the directory, which causes libprojfs to call into the C# code, which then calls down here again to a method like `WritePlaceholderDirectory()`.  ([Great catch](https://github.com/github/VFSForGit/pull/24#issuecomment-495061730), by the way!)

It still seems like we might have a some races (or at least some failure cases) between the time libprojfs gets things started in the `this.projfs.Start()` call below, and when we set our `this.projfs` variable above, or before FUSE has got everything up and working.

One idea I had looking at this was whether we could leave this method largely the same -- maybe just reverse the two lines:
```
            this.virtualizationRoot = virtualizationRootFullPath;
            this.projfs = fs;
```

and then, in the methods like `DeleteFile()` and `WritePlaceholderDirectory()` and such, add a check at the top which just returns `Result.EDriverNotLoaded` unless `this.projfs` is non-null.  And in our `StopVirtualizationInstance()` method below, reverse the lines so it goes:
```
            ProjFS fs = this.projfs;
            this.projfs = null;
            fs.Stop();
```

This does assume some atomicity on the part of setting C# member variables, and I'm not 100% sure how that works; maybe there's something that's needed from the `Interlocked` class?

Reply:
> It still seems like we might have a some races (or at least some failure cases) between the time libprojfs gets things started in the `this.projfs.Start()` call below, and when we set our `this.projfs` variable above, or before FUSE has got everything up and working.

I'm having trouble understanding what you're getting at here — where does the race occur? If we only call `Start()` after `this.projfs` is assigned, no callbacks can occur yet, as they only come from an active FUSE loop.

> ```
>             ProjFS fs = this.projfs;
>             this.projfs = null;
>             fs.Stop();
> ```

I'm not sure why this should be necessary either — once `Stop()` returns, every part of `libprojfs`'s working set has been stopped, deallocated, threads joined, etc. The `struct projfs` has even been freed, so if there was still something active, it'd likely segv way before it got to C#.

re: checks for `this.projfs`'s nullness, it would certainly mean these exceptions couldn't happen, but I also fear that kind of change in case it obfuscates something. If possible I'd like to design it so we don't need that kind of safety rail.

I feel like I'm missing something, so please let me know! In short what I'd like to ascertain is: where do you see a  race potentially occurring? What does "before FUSE has got everything up and working" mean in this context?

(Thank you for your thoughts here! It's this kind of icky edge case that I think requires the most attention and discussion.)

Reply:
Apologies -- you're seem absolutely right, and what I wrote should be largely dismissed as Friday afternoon nonsense.

Thinking this through as best I can over the weekend, I don't see any issues here.  The problems you discovered with threads accessing `this.projfs` and finding `null` all seem to be covered by the changes here, and those threads are presumably the ones started by libfuse's `mt` loop code, which as you say, should all be joined by the time `Stop()` returns, so there's no way they could then hit a `null` `this.projfs` in a subsequent callback.

So, apologies again for the brain cloud!

Reply:
> So, apologies again for the brain cloud!

Not at all! This is tricky stuff.